File: infmax.c
#include "poly.h"

NODE n[N];
GLOBAL g;

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// making a list of nodes that can be infected or recovered (very naive)

void get_infect_reco (int *infectables, int *ninfectables,
		int *recoverables, int *nrecoverables, int *infways) {
	int i, me, you;

	for (i = 0; i < g.n; i++) infways[i] = 0;

	for (me = 0; me < g.n; me++) {
		if (n[me].state == S) {
			for (i = 0; i < n[me].deg; i++) {
				you = n[me].nb[i];
				if (n[you].state == I) infways[me]++;
			}
			if (infways[me] > 0) infectables[(*ninfectables)++] = me;
		} else {
			if (n[me].state == I) recoverables[(*nrecoverables)++] = me;
		}
	}
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

void stepdown (fmpz_poly_t wnum0, fmpz_poly_t wden0) {
	int i, si = 0, you, infways[N];
	int infectables[N], ninfectables = 0, recoverables[N], nrecoverables = 0;
	fmpz_poly_t wnum, wden, num, den, a, b, c, d;

	fmpz_poly_init(wnum);
	fmpz_poly_init(wden);
	fmpz_poly_init(num);
	fmpz_poly_init(den);
	fmpz_poly_init(a);
	fmpz_poly_init(b);
	fmpz_poly_init(c);
	fmpz_poly_init(d);

	fmpz_poly_set(wnum, wnum0);
	fmpz_poly_set(wden, wden0);

	get_infect_reco(infectables, &ninfectables, recoverables, &nrecoverables, infways);
	
	for (i = 0; i < ninfectables; i++) si += infways[infectables[i]];

	if (si == 0) {
		fmpz_poly_mul(a, g.onum, wden);
		fmpz_poly_mul(b, g.oden, wnum);
		fmpz_poly_scalar_mul_ui(d, b, (slong) obsize());
		fmpz_poly_mul(c, g.oden, wden);
		fmpz_poly_set(g.oden, c);
		fmpz_poly_add(g.onum, a, d);

		simplify(&g.onum, &g.oden);

		goto CLEAR_EXIT;
	}

	fmpz_poly_zero(den);
	fmpz_poly_set_coeff_ui(den, 0, (unsigned long) nrecoverables);
	fmpz_poly_set_coeff_ui(den, 1, (unsigned long) si);

	for (i = 0; i < ninfectables; i++) {
		you = infectables[i];

		n[you].state = I;

		fmpz_poly_zero(num);
		fmpz_poly_set_coeff_ui(num, 1, (unsigned long) infways[you]);
		fmpz_poly_mul(a, wnum, num);
		fmpz_poly_mul(b, wden, den);

		stepdown(a, b);

		n[you].state = S;
	}
	
	for (i = 0; i < nrecoverables; i++) {
		you = recoverables[i];

		n[you].state = R;

		fmpz_poly_mul(a, wden, den);

		stepdown(wnum, a);

		n[you].state = I;
	}

	CLEAR_EXIT:

	fmpz_poly_clear(wnum);
	fmpz_poly_clear(wden);
	fmpz_poly_clear(num);
	fmpz_poly_clear(den);
	fmpz_poly_clear(a);
	fmpz_poly_clear(b);
	fmpz_poly_clear(c);
	fmpz_poly_clear(d);
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// main function handling input

int main (int argc, char *argv[]) {
	int i, j;
	char s1[OSIZE], s2[OSIZE], s3[OSIZE];
	fmpz_poly_t wnum, wden;

	if (argc < 3) {
		fprintf(stderr, "usage: ./infmax [# links] [links] <seeds>\n");
		return 1;
	}

	s1[0] = s2[0] = s3[0] = '\0';
	fmpz_poly_init(wnum);
	fmpz_poly_init(wden);
	fmpz_poly_init(g.onum);
	fmpz_poly_init(g.oden);

	// reading and setting up the network
	g.nl = atoi(argv[1]);
	g.n = 0;
	for (i = 0; i < N; i++) {
		n[i].deg = 0;
		n[i].state = S;
	}

	for (i = 0; i < g.nl; i++)
		add_edge(atoi(argv[2 + 2 * i]), atoi(argv[3 + 2 * i]));

	// setting the seeds
	for (i = 2 + 2 * g.nl, j = 0; i < argc; i++, j++) {
		n[atoi(argv[i])].state = I;   
		if (j > 0) strcat(s1, " ");
		strcat(s1, argv[i]);
	}

	fmpz_poly_one(wnum);
	fmpz_poly_one(wden);
	fmpz_poly_zero(g.onum);
	fmpz_poly_one(g.oden);

	stepdown(wnum, wden);
	simplify(&g.onum, &g.oden);
	
	strc(fmpz_poly_get_str_pretty(g.onum, "x"), s2);
	strc(fmpz_poly_get_str_pretty(g.oden, "x"), s3);
	printf("%s, (%s)/(%s)\n", s1, s2, s3);

	fmpz_poly_clear(wnum);
	fmpz_poly_clear(wden);
	fmpz_poly_clear(g.onum);
	fmpz_poly_clear(g.oden);

	return 0;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

File: infmax.py
# this code was written by Petter Holme spring / summer 2017
# it calculates the outbreak size of the SIR model given sets of seed nodes (i.e. influence maximization)
# it could be ran as
#   python infmax.py [# links] [links] <seeds>
# where [links] list the pairs of nodes connected in the network
# so for a triangle, and one seed node, one can run it as:
#   python infmax.py 3 0 1 1 2 2 0 0

import networkx as nx
from sys import argv
from sympy.abc import x
from sympy import Poly
from copy import deepcopy

# from gc import collect

#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #
# get the nodes that could be infected, or recovered

def get_infect_reco ():
	global G

	r = []
	for u,v in G.edges_iter():
		if G.node[u]['state'] == 'S' and G.node[v]['state'] == 'I':
			r.append(u)
		elif G.node[u]['state'] == 'I' and G.node[v]['state'] == 'S':
			r.append(v)
	
	infect = {}
	for rr in r:
		if rr in infect:
			infect[rr] += 1
		else:
			infect[rr] = 1

	recoverables = [v for v in G.nodes() if G.node[v]['state'] == 'I']
	return infect, recoverables


#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #

def obsize (G):
	n = 0
	for v in G.nodes():
		if G.node[v]['state'] == 'S':
			n += 1
		
	return G.number_of_nodes() - n

#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #
# stepping down in the tree of states, the path to the root is an infection chain

def stepdown (wnum0,wden0): # num means numerator, den denominator
	global G, onum, oden

	wnum = wnum0.mul_ground(1)
	wden = wden0.mul_ground(1)

	infectables, recoverables = get_infect_reco()
	si = sum(infectables.values())

	if si == 0: # if there are no nodes that can be infected, we can exit
		# this calculates onum/oden += wnum/wden * obsize(G)
		a = onum.mul(wden)
		b = wnum.mul(oden).mul_ground(obsize(G))
		c = oden.mul(wden)
		d = a.add(b)
		
		a = d.gcd(c)
		onum,no = d.div(a)
		oden,no = c.div(a)
		# collect() # if program takes too much memory 

		return

	sr = len(recoverables)
	# calculating the denominator for calculations below
	den = Poly(si * x + sr, x)

	for you, num in sorted(infectables.iteritems()):
		G.node[you]['state'] = 'I'
		# stepdown with a weight: num * x * [wnum/wden] / den
		
		stepdown(wnum.mul(Poly(num * x,x)),wden.mul(den))
		
		G.node[you]['state'] = 'S'
	for you in sorted(recoverables):
		G.node[you]['state'] = 'R'
		# stepdown with a weight: [wnum/wden] / den
		
		stepdown(wnum,wden.mul(den))
		
		G.node[you]['state'] = 'I'
	# collect() # if program takes too much memory

#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #

if __name__ == "__main__":
	global G, onum, oden

	if len(argv) < 2:
		print 'usage python infmax.py [# links] [links] <seeds>'
		exit()

	G0 = nx.Graph()

	nl = int(argv[1])

	for i in range(nl):
		G0.add_edge(argv[2 + 2 * i], argv[3 + 2 * i])

	for v in G0.nodes():
		G0.node[v]['state'] = 'S'

	s = ''
	for i in range(2 + 2 * nl,len(argv)):
		me = argv[i]
		s += me + ' '
		if me not in G0.nodes():
			print me, 'not in G'
			exit()
		G0.node[me]['state'] = 'I'

	#G = nx.convert_node_labels_to_integers(G0,label_attribute='id')
	G = deepcopy(G0)

	onum = Poly(0,x)
	oden = Poly(1,x)
	stepdown(Poly(1,x),Poly(1,x))
	
	a = onum.gcd(oden)
	onum, no = onum.div(a)
	oden, no = oden.div(a)

	# the output format is: [active nodes (separated by blanks)], [solution polynomial]
	print s.strip() + ', (' + str(onum.as_expr()) + ')/(' + str(oden.as_expr()) + ')'

#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #
